### ISSUE TITLE
Dask kernel dies when fusing large images

### ISSUE THREAD

**Issue by dpshepherd** [multiview-stitcher/multiview-stitcher#25]
We've been running into an issue where our Ubuntu 22.04 machine kills Dask when trying to fuse large images. We seem to have temporarily solved it by stopping the out-of-memory killer, but this introduces its own problems.

Would you be up for a call / pair coding session at some point to explore alternative frameworks to Dask? 

We've had good luck for larger than memory operations combining lightweight tiling via [Ryomen](https://github.com/buswinka/ryomen) and Tensorstore, but there are lots of options here.

**Comment by m-albert** [comment]
Hi, good to know you're running into these problems. Did you start experiencing this with a certain version of multiview-stitcher? This is just to make sure there's no significant regression of the package in that regard.

In any case I can confirm that for very large images the computation of the dask graph becomes very inefficient and/or the scheduler (or workers) get killed by the OS. However, I need to do some more debugging to identify the bottleneck more clearly.

Currently there's potential to significantly optimize the complexity of the dask graphs produced by `multiview_stitcher.fusion.fuse`. The current code is much too generous in adding tasks. Ideally I'd like to boil it down to a single call of `map_blocks`.

> Would you be up for a call / pair coding session at some point to explore alternative frameworks to Dask?
> We've had good luck for larger than memory operations combining lightweight tiling via [Ryomen](https://github.com/buswinka/ryomen) and Tensorstore, but there are lots of options here.

I'd be very open and curious to discuss about how to improve performance and support larger data (also adding [`xarray-beam`](https://xarray-beam.readthedocs.io/en/latest/index.html) for reference here).

Very happy to have a chat about this, e.g. Wed-Fri this week or Tue-Thu next week. Generally the European afternoon would suit me well (feel free to suggest other times though).

**Comment by dpshepherd** [comment]
> Did you start experiencing this with a certain version of multiview-stitcher?

This has always been an issue with multiview-stitcher, it just took us a long time to narrow it down the Dask scheduler.

> Very happy to have a chat about this, e.g. Wed-Fri this week or Tue-Thu next week. Generally the European afternoon would suit me well (feel free to suggest other times though).

Next week is better for me. Late afternoon is usually early morning here in Arizona, that is fine for me.

**Comment by m-albert** [comment]
> This has always been an issue with multiview-stitcher, it just took us a long time to narrow it down the Dask scheduler. 

That's good to know. I'm currently working on cleaning up / improving the way fused dask arrays are constructed, which will streamline and strongly reduce the amount of total tasks and make the memory requirement to calculate each output chunk more transparent. This will hopefully bump the dataset sizes manageable with the current setup. 

> Next week is better for me. Late afternoon is usually early morning here in Arizona, that is fine for me. 

Great! Feel free to suggest a day and time. You can also contact me over marvin.albert@gmail.com

**Comment by m-albert** [comment]
Hi @dpshepherd, here's a quick update before I leave for holidays next week (will be available again the week after):

I've been working on improving the fusion process and opened a PR [here](https://github.com/multiview-stitcher/multiview-stitcher/pull/26). See also the PR description.

It'd be interesting to know whether this performs better on your data (branch simplified_fusion).

**Comment by m-albert** [comment]
Quick update: `multiview-stitcher` version 0.1.14 contains changes from https://github.com/multiview-stitcher/multiview-stitcher/pull/26 which improve memory requirements significantly. I recommend using the threads scheduler.

**Comment by dpshepherd** [comment]
Sorry for dropping this, I had some travel, got sick, and then the semester started.

Thanks for your work! We will test it out on some datasets and get back you. Much appreciated!

**Comment by m-albert** [comment]
No worries, thanks a lot for writing!

Looking forward to hearing whether the changes in multiview-stitcher >= 0.1.14 improve things for you.

**Comment by dpshepherd** [comment]
For one channel, things definitely seem improved. For multiplexed imaging (e.g. 18 channels of iterative labeling), the memory still runs away and then the kernel dies. Is there a way we should adapt how we call the fusion function for many channels?

**Comment by m-albert** [comment]
Thanks for the feedback. Interesting, so if I understand you correctly, things work well with a single channel but fail for many channels.

I'm wondering whether there's a problem with using `xr.combine_by_coords` on large dask backed `xr.DataArray` here: https://github.com/multiview-stitcher/multiview-stitcher/blob/ac2e7cfbf848b5445ce1a04abb03fc372f821e7a/src/multiview_stitcher/fusion.py#L450

Will look into this in the next days. What array sizes (18, ...?) are you approx. working with?

> Is there a way we should adapt how we call the fusion function for many channels?

No, the API of `fusion.fuse` didn't change. However if things currently work well with a single channel you could fuse the channels independently. E.g.:

```python
fused_GFP = [fusion.fuse(
    [sim.sel({'c': 'GFP'}) for sim in sims],
    transform_key='registered',
    output_spacing={dim: 10 for dim in ['z', 'y', 'x']},
    output_chunksize=128,
    )
```

**Comment by dpshepherd** [comment]
> Thanks for the feedback. Interesting, so if I understand you correctly, things work well with a single channel but fail for many channels.

Correct. I believe the individual FOVs are something like `(18,65,2048,2048)` in `czyx` order. And there on the order of 50-100 FOVs.

**Comment by dpshepherd** [comment]
Saving as independent channels works if we downsample the fused image output spacing. Strangely, if we use the the full resolution data it saves all of the chunks, but everything is empty.

The use case is we are now trying to register and fusion ~2000-3000 individual 2D images from high-resolution whole slide imaging. In addition to the above, there appear to be many slightly misaligned tiles in the final fusion. We'll narrow down to a working example and put some data/code online for you.

**Comment by m-albert** [comment]
> Saving as independent channels works if we downsample the fused image output spacing. Strangely, if we use the the full resolution data it saves all of the chunks, but everything is empty.

So it runs through but writes empty chunks to disk. That's strange. Related to that, I'm thinking it'd probably be a good idea to add a function e.g. `fusion.fuse_large` which sequentially fuses and writes manageable subsets of the array to be fused. Would that make sense to you?

> The use case is we are now trying to register and fusion ~2000-3000 individual 2D images from high-resolution whole slide imaging. In addition to the above, there appear to be many slightly misaligned tiles in the final fusion. We'll narrow down to a working example and put some data/code online for you.

Good to know. An example to debug this would be great 👌. Thanks!

**Comment by m-albert** [comment]
This is fixed! See https://github.com/multiview-stitcher/multiview-stitcher/issues/94#issuecomment-3409877549.
